### PR TITLE
Annotation tool

### PR DESCRIPTION
**Pull request by simra**

### CODE REVIEW

File: src/annotation/app.py [hunk 1]
```diff
@@ -85,6 +87,30 @@ def post_annotation():
     return jsonify({"message": "Annotation posted", "id": cur.lastrowid})
 
 
+@app.route("/replace_annotation/<int:anno_id>", methods=['POST'])
```
Comment (monkeypants) [inline src/annotation/app.py:95]: `POST {new data} /replace_annotation/{id}` seems a bit unnatural (why not `PUT {new data} /annotations/{id}`)?

File: src/annotation/app.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+# TODO: add casing detection model and pre-compute detections
+# TODO: mouseover polygon color change
+# TODO: better db representation of metadata, geometry
+# TODO: migrate to jquery/modern widget framework
+# TODO: deal with unusual image shapes, enable zoom, panning, etc. tiling very large images
+# TODO: double-check replace events are updating the annotation id correctly.
+# TODO: adding more images
+# TODO: host db online/ migrate to modern/robust db.
+# TODO: e2e image processing pipeline/user experience
+# TODO: proper RESTful API
```
Comment (monkeypants) [inline src/annotation/app.py:28]: this list of TODOs might make their way into tickets eventually :)

File: src/annotation/app.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+@app.route("/")
+def home():
+    return redirect('/annotate/')
+
+
+@app.route("/img/<id>")
+def img(id):
+    try:
+        cur = get_db().cursor()
+        cur.execute("SELECT * FROM images WHERE img_id=={}".format(id))
```
Comment (monkeypants) [inline src/annotation/app.py:47]: per #11, these embedded SQL statements present an injection surface

File: src/annotation/app.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    for row in cur.fetchall():
+        row = dict(zip(columns, row))
+        row['geometry'] = json.loads(row['geometry'])
+        row['metadata'] = json.loads(row['metadata'])
+        results.append(row)
+
+    return jsonify(results)
+
+
+@app.route("/post_annotation", methods=['POST'])
```
Comment (monkeypants) [inline src/annotation/app.py:73]: idiomatic REST would be `POST /annotations`

File: src/annotation/app.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    cur.execute(cmd1)
+    cur.execute(cmd)
+    con.commit()
+    cur.execute('SELECT COUNT(*) AS c FROM annotations')
+    result = next(cur, [None])['c']
+    logging.info('found {} rows'.format(result))
+    return jsonify({"message": "Annotation posted", "id": cur.lastrowid})
+
+
+@app.route("/delete_annotation/<int:anno_id>", methods=['DELETE'])
```
Comment (monkeypants) [inline src/annotation/app.py:121]: also `DELETE /annotations/{id}`

File: src/annotation/app.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+
+@app.route("/annotate/")
+@app.route("/annotate/<int:id>")
+def annotate(id=None):
+    show_annot = request.args.get('show_annot')
+    if id is None:
+        cur = get_db().cursor()
+        if show_annot:
+            cur.execute("SELECT MIN(images.img_id) AS id from images")
```
Comment (monkeypants) [inline src/annotation/app.py:143]: bad things happen if the number of images is 0. That's probably not a big problem.

File: src/annotation/app.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    con = get_db()
+    cur = con.cursor()
+    cmd = "DELETE FROM annotations WHERE anno_id={}".format(anno_id)
+    logging.info(cmd)
+    cur.execute(cmd)
+    con.commit()
+    return jsonify(success=True)
+
+
+@app.route("/annotate/")
```
Comment (monkeypants) [inline src/annotation/app.py:136]: idiomatic rest would have a "plural noun" here.

**Review by monkeypants — APPROVED** [review]
it works and doesn't break anything, so please merge it :)

The suggestions about changing the paths in the REST API are a bit invasive (will break the UI) so that might be best as a separate refactor PR. If you agree, I can have a go at that in the next few days.

File: src/annotation/app.py (hunk 1 above)
Comment (simra) [reply src/annotation/app.py:95]: Thanks for reviewing. Yes, if you don't mind cleaning up the API that would be great.  I'll go ahead and merge.

**Pull request merged**